packages/components/src/components/Calculator/useSavedCalculatorState.tsx

```diff
@@ -0,0 +1,56 @@
+import { useCallback, useMemo, useState } from "react";
+
+import { useViewerContext } from "../SquiggleViewer/ViewerProvider.js";
+import { SqCalculatorValueWithContext } from "./types.js";
+import { SqValueResult } from "./types.js";
+import { FormShape } from "./types.js";
+import { InputValues } from "./types.js";
+
+export type CalculatorState = {
+  hashString: string;
+  formValues?: FormShape;
+  inputValues?: InputValues;
+  fnValue?: SqValueResult;
+};
+
+// Takes a calculator value; returns the cache from ViewerContext for this calculator and a function for updating the cache.
+// Note that the returned cached value is never updated! It's write-only after the initial load.
+export function useSavedCalculatorState(
+  calculatorValue: SqCalculatorValueWithContext
+) {
+  const path = useMemo(() => calculatorValue.context.path, [calculatorValue]);
+
+  const { getLocalItemState, dispatch: viewerContextDispatch } =
+    useViewerContext();
+
+  // Load cache just once on initial render.
+  // After the initial load, this component owns the state and pushes it to ViewerContext.
+  const [cachedState] = useState<CalculatorState | undefined>(() => {
+    const itemState = getLocalItemState({ path });
+
+    const sameCalculatorCacheExists =
+      itemState.calculator &&
+      itemState.calculator.hashString === calculatorValue.value.hashString;
+
+    if (sameCalculatorCacheExists) {
+      return itemState.calculator;
+    } else {
+      return undefined;
+    }
+  });
+
+  const updateCachedState = useCallback(
+    (state: CalculatorState) => {
+      viewerContextDispatch({
+        type: "CALCULATOR_UPDATE",
```

Comment:
Minor, but it could be nice to rename this to "CALCULATOR_CACHE_UPDATE"

Reply:
In my new code, I left a few comments that treated it as cache, but it's not just for caching as I remembered later: we also rely on this object for tracking collapse states.

Maybe I should rename the `cachedState` because of this.

(Alternatively, maybe CalculatorState should include the full ViewerContext for the calculator result, so that we could mount a separate SquiggleViewer to render the function result. But that's another big refactoring, and I'm not even sure if it's a good idea or not.)